File: 237G_Computational/Projects/Project1/Python/Question1.py
import statistics
from numpy import random

def question1(n, seed):
    randnum = generate_uniform_dist(n, seed)
    mean = statistics.mean(randnum)
    stddev = statistics.stdev(randnum)
    print("Mean of LGM is {0:.4f}".format(mean))
    print("Standard Deviation of LGM is {0:.4f}".format(stddev))

    inbuilt_random = random.uniform(0,1,n)
    mean_inbuilt = statistics.mean(inbuilt_random)
    stddev_inbuilt = statistics.mean(inbuilt_random)
    print("Mean of Inbuilt is {0:.4f}".format(mean_inbuilt))
    print("Standard Deviation of Inbuilt is {0:.4f}".format(stddev_inbuilt))


def generate_uniform_dist(N, seed):
    m = 2 ** 31 - 1
    a = 7 ** 5
    b = 0

    randnum = [seed]

    for count in range(1, N+1):
        temp = a * (randnum[count - 1]) + b
        randnum.append(temp % m)

    del randnum[0]
    randnum = [(val + 0.5) / m for val in randnum]

    return randnum


def generate_uniform_dist_prev(N, prev_val):
    m = 2 ** 31 - 1
    seed = prev_val * m - 0.5
    return generate_uniform_dist(N,seed)

File: 237G_Computational/Projects/Project1/ComputationalFinance_Project1/Project1/Question1.cpp
#include "Question1.h"
#include <random>
using namespace std;

const unsigned long long m = pow(2, 31) - 1;
const long a = pow(7, 5);
const int b = 0;

void Question1(int N, double seed)
{
	vector<double> randNum = GenerateUniformDist(N, seed);

	double mean_lgm = mean(randNum);
	double sd_lgm = stddev(randNum);
	cout << endl << "QUESTION 1" << endl;
	cout << "The Mean of LGM is " << mean_lgm << endl;
	cout << "The Standard Deviation of LGM is " << sd_lgm << endl;

	//1b - Use C++ generator
	default_random_engine generator;
	uniform_real_distribution<double> distribution(0.0, 1.0);
	vector<double> randNum_inbuilt;
	for (int count = 0; count < N; count++)
	{
		randNum_inbuilt.push_back(distribution(generator));
	}
	double mean_inbuilt = mean(randNum_inbuilt);
	double sd_inbuilt = stddev(randNum_inbuilt);
	cout << endl << "The Mean of Inbuilt is " << mean_inbuilt << endl;
	cout << "The Standard Deviation of Inbuilt is " << sd_inbuilt << endl;

	cout << endl << "The mean and standard deviation between the two methods are very similar." << endl;
	cout << "The small difference might be due to the choice of the seed and the default random generator might not be using the same parameters as the LGM method"<<endl;
}

vector<double> GenerateUniformDist(int N,double seed)
{
	//1a - random number generator
	vector<double> randNum;
	//add seed. will remove it later
	randNum.push_back(seed);
	for (int count = 1; count <= N; count++)
	{
		unsigned long long temp = a*(randNum[count - 1]) + b;
		randNum.push_back(temp % m);
	}
	randNum.erase(randNum.begin());
	//Convert to U(0,1)
	for (int count = 0; count < N; count++)
	{
		randNum[count] = (randNum[count] + 0.5) / m;
	}

	return randNum;
}

vector<double> GenerateUniformDist_UsingPrev(int N, double prevVal)
{
	double seed = prevVal*m - 0.5;
	return GenerateUniformDist(N, seed);
}

